modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/rel/IndexScanNodeTest.java

```diff
@@ -0,0 +1,288 @@
+
+        Comparator<Object[]> cmp = (a, b) -> {
+            Comparable o1 = (Comparable<?>) a[0];
+            Comparable o2 = (Comparable<?>) b[0];
+            return o1.compareTo(o2);
+        };
+
+        Tester tester = new Tester();
+
+        IndexScanNode<Object[]> node = tester.createSortedIndex(index, scannableTable, cmp);
+        List<Object[]> result = tester.execute(node);
+
+        tester.expectResult(result, new Object[]{1}, new Object[]{2}, new Object[]{4}, new Object[]{5});
+    }
+
+    /**
+     * Hash index lookup execution.
+     */
+    @Test
+    public void testHashIndex() {
+        IgniteIndex index = newHashIndex();
+
+        TestScannableTable<Object[]> scannableTable = new TestScannableTable<>();
+        scannableTable.setPartitionData(0, new Object[]{2}, new Object[]{1});
+        scannableTable.setPartitionData(2, new Object[]{0});
+
+        Tester tester = new Tester();
+
+        IndexScanNode<Object[]> node = tester.createHashIndex(index, scannableTable);
+        List<Object[]> result = tester.execute(node);
+
+        tester.expectResult(result, new Object[]{2}, new Object[]{1}, new Object[]{0});
+    }
+
+    private static IgniteIndex newHashIndex() {
+        IndexDescriptor descriptor = new IndexDescriptor("IDX", List.of("C1"));
+        HashIndex index = new HashIndex(1, Mockito.mock(InternalTable.class), descriptor);
+
+        return new IgniteIndex(index);
+    }
+
+    private static IgniteIndex newSortedIndex() {
+        List<String> columnNames = List.of("C1");
+        List<ColumnCollation> columnCollations = List.of(ColumnCollation.ASC_NULLS_LAST);
+        SortedIndexDescriptor descriptor = new SortedIndexDescriptor("IDX", columnNames, columnCollations);
+        SortedIndexImpl index = new SortedIndexImpl(1, Mockito.mock(InternalTable.class), descriptor);
+
+        return new IgniteIndex(index);
+    }
+
+    private class Tester {
+
+        private final ExecutionContext<Object[]> ctx;
+
+        Tester() {
+            this.ctx = executionContext();
+        }
+
+        IndexScanNode<Object[]> createSortedIndex(IgniteIndex index, TestScannableTable<?> scannableTable, Comparator<Object[]> cmp) {
+            return createIndexNode(ctx, index, scannableTable, cmp);
+        }
+
+        IndexScanNode<Object[]> createHashIndex(IgniteIndex index, TestScannableTable<?> scannableTable) {
+            return createIndexNode(ctx, index, scannableTable, null);
+        }
+
+        List<Object[]> execute(IndexScanNode<Object[]> indexNode) {
+            RootNode<Object[]> root = new RootNode<>(ctx);
+
+            root.register(indexNode);
+
+            List<Object[]> actual = new ArrayList<>();
+            while (root.hasNext()) {
+                Object[] row = root.next();
+                actual.add(row);
+            }
+
+            root.close();
+
+            return actual;
+        }
+
+        void expectResult(List<Object[]> actual, Object[]... expected) {
```

Comment:
```suggestion
        static void validateResult(List<Object[]> actual, Object[]... expected) {
```
Also, I think `actual` and `expected` should be of the same type: either both List or arrays.

Reply:
Fixed.